### ISSUE TITLE
`suboptimal_pattern` false positive

### ISSUE THREAD

**Issue by smoelius** [trailofbits/dylint#1093]
On the commit just prior to this one https://github.com/trailofbits/cargo-unmaintained/commit/20292f88e6958aedbcf3cc20adff088418fb5b68, `suboptimal_pattern` produces a bad suggestion.

I haven't figured out why.

**Comment by smoelius** [comment]
Here is the relevant code:
```rs
    display_advisory_outcomes(
        &advisory_outcomes
            .iter()
            .map(|(advisory, outcome)| {
                let url = advisory_url(advisory);
                (&advisory.metadata.package, url, *outcome)
            })
            .collect::<Vec<_>>(),
    );
```
Here is the warning:
```
warning: could destructure reference
   --> src/bin/rustsec_advisories.rs:123:30
    |
123 |             .map(|(advisory, outcome)| {
    |                              ^^^^^^^ help: use: `&outcome`
    |
    = note: `#[warn(suboptimal_pattern)]` on by default
```
And here is the error that results:
```
error[E0308]: mismatched types
   --> src/bin/rustsec_advisories.rs:123:30
    |
123 |             .map(|(advisory, &outcome)| {
    |                   -----------^^^^^^^^-
    |                   |          |
    |                   |          expected `Outcome<Reason>`, found `&_`
    |                   expected due to this
    |
    = note:   expected enum `Outcome<Reason>`
            found reference `&_`
help: consider removing `&` from the pattern
    |
123 -             .map(|(advisory, &outcome)| {
123 +             .map(|(advisory, outcome)| {
    |
```

**Comment by smoelius** [comment]
Fixed by #1512
